instrumentation/opentelemetry-instrumentation-sklearn/src/opentelemetry/instrumentation/sklearn/__init__.py

```diff
@@ -0,0 +1,757 @@
+# limitations under the License.
+"""
+The integration with sklearn supports the scikit-learn compatible libraries,
+it can be enabled by using ``SklearnInstrumentor``.
+
+.. sklearn: https://github.com/scikit-learn/scikit-learn
+
+Usage
+-----
+
+Package instrumentation example:
+
+.. code-block:: python
+
+    from opentelemetry.instrumentation.sklearn import SklearnInstrumentor
+
+    # instrument the sklearn library
+    SklearnInstrumentor().instrument()
+
+    # instrument sklearn and other libraries
+    SklearnInstrumentor(
+        packages=["sklearn", "lightgbm", "xgboost"]
+    ).instrument()
+
+
+Model intrumentation example:
+
+.. code-block:: python
+
+    from opentelemetry.instrumentation.sklearn import SklearnInstrumentor
+    from sklearn.datasets import load_iris
+    from sklearn.ensemble import RandomForestClassifier
+    from sklearn.model_selection import train_test_split
+    from sklearn.pipeline import Pipeline
+
+    X, y = load_iris(return_X_y=True)
+    X_train, X_test, y_train, y_test = train_test_split(X, y)
+
+    model = Pipeline(
+        [
+            ("class", RandomForestClassifier(n_estimators=10)),
+        ]
+    )
+
+    model.fit(X_train, y_train)
+
+    SklearnInstrumentor().instrument_estimator(model)
+
+"""
+import logging
+import os
+from functools import wraps
+from importlib import import_module
+from inspect import isclass
+from pkgutil import iter_modules
+from typing import Callable, Dict, List, MutableMapping, Sequence, Type, Union
+
+from sklearn.base import BaseEstimator
+from sklearn.pipeline import FeatureUnion, Pipeline
+from sklearn.tree import BaseDecisionTree
+from sklearn.utils.metaestimators import _IffHasAttrDescriptor
+
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.instrumentation.sklearn.version import __version__
+from opentelemetry.trace import get_tracer
+from opentelemetry.util.types import Attributes
+
+logger = logging.getLogger(__name__)
+
+
+def implement_spans(
+    func: Callable,
+    estimator: Union[BaseEstimator, Type[BaseEstimator]],
+    attributes: Attributes = None,
+):
+    """Wrap the method call with a span.
+
+    Args:
+        func: A callable to be wrapped in a span
+        estimator: An instance or class of an estimator
+        attributes: Attributes to apply to the span
+
+    Returns:
+        The passed function wrapped in a span.
+    """
+    if isclass(estimator):
+        name = estimator.__name__
+    else:
+        name = estimator.__class__.__name__
+    logger.debug("Instrumenting: %s.%s", name, func.__name__)
+
+    attributes = attributes or {}
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        with get_tracer(__name__, __version__).start_as_current_span(
+            name="{cls}.{func}".format(cls=name, func=func.__name__),
+        ) as span:
+            if span.is_recording():
+                for key, val in attributes.items():
```

Comment:
tracing the call tree, it looks like this facility only supports static attributes to annotate an instrumentation.

I would wonder if it's valuable to add a callable edition, where it will pass in the object that is starting the operation, so additional information can be extracted (e.g. some parameterization of the model)?

It's not like this couldn't be added later, just a thought.